### PR TITLE
support awareness of in-flight nodes & startup taints

### LINKED ISSUE #628: Support extra taints not considered when scheduling Pods
**Tell us about your request**

As of today, if taints are defined in [`spec.Constraints.taints`](https://karpenter.sh/docs/reference/#karpenter.sh/v1alpha3.Constraints), the provisioner will not provision nodes for pods that do not have matching tolerations.

> Taints will be applied to every node launched by the Provisioner. If specified, the provisioner will not provision nodes for pods that do not have matching tolerations.

However, some cluster add-ons (usually CNI plugins) use temporary taints on new nodes to make sure no Pods are scheduled onto that nodes unless the CNI `DaemonSet` is ready on that node. Usually, those *temporary node startup taints* are then removed by the `DaemonSet` or by a corresponding controller. A prominent example is [Cilium](https://docs.cilium.io/en/stable/gettingstarted/k8s-install-default/) which uses the `node.cilium.io/agent-not-ready` taint. 

Adding toleration to all Pods for such *temporary node startup taints* would of course void their purpose and might lead to disfunctional Pods as they are scheduled and started before CNI was properly setup on that node. With its current capabilities karpenter is not powerful enough to support such a setup.

**What do you want us to build?**

Add something like `spec.Constraints.extraTaints`, which contains taints that must be applied to newly created nodes, but might later be removed by workloads and are ignored by the provisioner for scheduling purposes. Specifically, karpenter should support the following behaviour:

- provisioner adds all `extraTaints` to newly created nodes and then forgets about them (does not re-add them if they are later removed by a CNI controller).
- provisioner ignores `extraTaints` when making scheduling decisions

Of course, the proposed solution is just a (viable) placeholder and maybe somebody has a better ideas to resolve this problem, so feel free to propose other solutions.

**Are you currently working around this issue?**

Not using karpenter at all :-(  

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by tzneal**
**1. Issue, if available:**

Fixes #628 and #972

**2. Description of changes:**

- don't launch nodes if the pod can schedule on a recently launch
  node that hasn't become ready
- support startup taints

Co-authored-by: Bryan Stenson <bryan.stenson@okta.com>

**3. How was this change tested?**

Unit testing & scaling cluster on EKS. I tested with two EKS clusters, one using Cilium and one standard cluster.

**4. Does this change impact docs?**
- [X] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: website/content/en/preview/provisioner.md
Comment (BryanStenson-okta) [inline website/content/en/preview/provisioner.md:33]: ^determing^determining

File: pkg/apis/provisioning/v1alpha5/constraints.go
Comment (tzneal) [inline pkg/apis/provisioning/v1alpha5/constraints.go:38]: Clarify that these don't cause taints to be added to nodes.

File: pkg/controllers/provisioning/scheduling/inflightnode.go
Comment (tzneal) [inline pkg/controllers/provisioning/scheduling/inflightnode.go:55]: Remove this to eliminate the two different timeout values

File: pkg/controllers/provisioning/scheduling/inflightnode.go
Comment (tzneal) [inline pkg/controllers/provisioning/scheduling/inflightnode.go:89]: Try to track this in cluster state.

File: pkg/controllers/provisioning/scheduling/node.go
Comment (tzneal) [inline pkg/controllers/provisioning/scheduling/node.go:39]: Rename to startupTolerations

File: pkg/controllers/provisioning/scheduling/scheduler.go
Comment (tzneal) [inline pkg/controllers/provisioning/scheduling/scheduler.go:62]: Look at the deleted timestamp.

Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/scheduler.go:62]: Already handled by cluster state, just adding a test.

**Comment by tzneal** [comment]
Tested and I am able to launch the cilium connectivity test on a single node cluster. This leads to unschedulable pods for the multi-node tests, so Karpenter launches a second node and doesn't bind the pods to the node due to the startup taint.  Eventually cilium becomes ready, removes the startup taint and kube-scheduler binds the pods.

```
NAMESPACE     NAME                                                     READY   STATUS    RESTARTS   AGE     IP                NODE                                           NOMINATED NODE   READINESS GATES
cilium-test   echo-a-c4cdff77c-jr6bh                                   1/1     Running   0          4m7s    192.168.156.116   ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
cilium-test   echo-b-598c78b9fc-h458m                                  1/1     Running   0          4m6s    192.168.34.22     ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
cilium-test   echo-b-host-5556f9488-rxdkk                              1/1     Running   0          4m5s    192.168.55.185    ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
cilium-test   host-to-b-multi-node-clusterip-7f8c9699d6-ppc62          1/1     Running   0          4m1s    192.168.11.65     ip-192-168-11-65.us-west-2.compute.internal    <none>           <none>
cilium-test   host-to-b-multi-node-headless-7df4d6fdb-zkfc4            1/1     Running   0          4m1s    192.168.11.65     ip-192-168-11-65.us-west-2.compute.internal    <none>           <none>
cilium-test   pod-to-a-679f686cb-frndq                                 1/1     Running   0          4m5s    192.168.50.51     ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
cilium-test   pod-to-a-allowed-cnp-54755cc9c6-7tr2v                    1/1     Running   0          4m3s    192.168.158.115   ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
cilium-test   pod-to-a-denied-cnp-7bfb7d69b8-kv4wf                     1/1     Running   0          4m4s    192.168.138.176   ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
cilium-test   pod-to-b-intra-node-nodeport-6bd8d9468d-4cncq            1/1     Running   0          3m59s   192.168.158.6     ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
cilium-test   pod-to-b-multi-node-clusterip-7984ccf8c6-v9vm9           1/1     Running   0          4m2s    192.168.100.199   ip-192-168-11-65.us-west-2.compute.internal    <none>           <none>
cilium-test   pod-to-b-multi-node-headless-7fb5f5c84b-g4654            1/1     Running   0          4m2s    192.168.117.93    ip-192-168-11-65.us-west-2.compute.internal    <none>           <none>
cilium-test   pod-to-b-multi-node-nodeport-54c89d778-hhggm             1/1     Running   0          4m      192.168.12.141    ip-192-168-11-65.us-west-2.compute.internal    <none>           <none>
cilium-test   pod-to-external-1111-c98db84d4-6qdph                     1/1     Running   0          4m4s    192.168.59.118    ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
cilium-test   pod-to-external-fqdn-allow-google-cnp-5f55d8886b-846hw   1/1     Running   0          4m3s    192.168.62.239    ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
karpenter     karpenter-c4954dbf-wwmtl                                 2/2     Running   0          5m20s   192.168.153.126   ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
kube-system   cilium-52q78                                             1/1     Running   0          12h     192.168.55.185    ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
kube-system   cilium-node-init-9x2xg                                   1/1     Running   0          3m55s   192.168.11.65     ip-192-168-11-65.us-west-2.compute.internal    <none>           <none>
kube-system   cilium-node-init-z9vb7                                   1/1     Running   0          12h     192.168.55.185    ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
kube-system   cilium-operator-6689cf95b6-thpwb                         1/1     Running   0          12h     192.168.55.185    ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
kube-system   cilium-vb6l7                                             1/1     Running   0          3m13s   192.168.11.65     ip-192-168-11-65.us-west-2.compute.internal    <none>           <none>
kube-system   coredns-85d5b4454c-4p9s9                                 1/1     Running   0          13h     192.168.130.87    ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
kube-system   coredns-85d5b4454c-rq4vr                                 1/1     Running   0          34h     192.168.43.71     ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
kube-system   kube-proxy-5lrqf                                         1/1     Running   0          34h     192.168.55.185    ip-192-168-55-185.us-west-2.compute.internal   <none>           <none>
kube-system   kube-proxy-x92cm                                         1/1     Running   0          3m13s   192.168.11.65     ip-192-168-11-65.us-west-2.compute.internal    <none>           <none>
```

File: pkg/controllers/provisioning/provisioner.go
Comment (suket22) [inline pkg/controllers/provisioning/provisioner.go:122]: If we do do this, should we have a debug log describing a bind to an in-flight node? It might surprise users that are expecting a new node to launch but none were

File: pkg/controllers/provisioning/scheduling/scheduler.go
Comment (suket22) [inline pkg/controllers/provisioning/scheduling/scheduler.go:74]: If an instance doesn't launch because of an ICE or because the EC2 service is unschedulable, wouldn't we end up considering this node object as `inflight` ? Should we only be doing this once we've kicked off the node, rather than after we've completed successful scheduling?

File: pkg/apis/provisioning/v1alpha5/constraints.go
Comment (tzneal) [reply pkg/apis/provisioning/v1alpha5/constraints.go:42]: I agree this was confusing.  I've left it named as 'StartupTaints' for now, but I've modified it so that it now actually causes the taints to be applied to the node.    You'll just list the taint as a startup taint, we put it on the node but ignore it for scheduling and assume someone else will remove it.

File: pkg/controllers/provisioning/provisioner.go
Comment (tzneal) [reply pkg/controllers/provisioning/provisioner.go:122]: I'm about to push up a new PR that starts to add support for recording events on the pods (at least an interface for it).  We needed something like that for unit tests anyway as we need to know if the scheduling decisions were correct, even without the pod binding.  There is a separate issue at #1584 to turn those into actual events on the pods.  It's more code and requires changing permissions, so I didn't include it in this PR.

File: pkg/controllers/provisioning/scheduling/scheduler.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/scheduler.go:74]: This shouldn't occur, we get a success from the instance launch before we create the node object.  The inflight nodes here are ones that were created in previous rounds of scheduling.

There is a scenario where a node fails to become ready and we think it will support pods.  It may prevent provisioning whatever excess capacity it has for 15 minutes, but it will resolve.  The same issue occurs currently where we bind the pods to the node that never comes up and they don't retry for 15 minutes.

File: pkg/controllers/provisioning/scheduling/inflightnode.go
Comment (bwagner5) [inline pkg/controllers/provisioning/scheduling/inflightnode.go:82]: any more thought on combining or at least re-using some of the logic in  `scheduling/node.go`?

**Review by bwagner5 — COMMENTED** [review]
overall, looks good! No major concerns, just a few minor comments.

File: charts/karpenter/crds/karpenter.sh_provisioners.yaml
Comment (tzneal) [reply charts/karpenter/crds/karpenter.sh_provisioners.yaml:129]: We don't, but it's part of the v1.Taint.  The same field is on each of our regular Taints. We would need to create our own type to remove it.

File: cmd/controller/main.go
Comment (tzneal) [reply cmd/controller/main.go:82]: I did this specifically so that everything that uses `recorder` would have to use the `events.Recorder` interface and we couldn't accidentally use the real type.

**Comment by ellistarn** [comment]
I'm wary of the long term impact of including startup taints in the api. It's discussed in a number of places, but I'll restate here. Upstream has a concept to address this use case (including the node restart case) called the NoAdmit taint effect. This allows scheduling but prevents the kubelet from running the pod, which is exactly what we need. 

I'd hate to include this in karpenters API as a point in time workaround for something that will eventually be fixed upstream. One option is to support this taint effect in karpenter before it's merged upstream and do the upstream work in parallel. 

There may be reasons to not do this, and we can always deprecate a field using an API bump, but I want to make sure we're rigorously evaluating our options for changes that expand the API.

**Comment by tzneal** [comment]
> I'm wary of the long term impact of including startup taints in the api. It's discussed in a number of places, but I'll restate here. Upstream has an unimplemented concept to address this use case (including the node restart case) called the NoAdmit taint effect. This allows scheduling but prevents the kubelet from running the pod, which is exactly what we need. One benefit of doing this upstream is that the change applies to all integrators like CAS, Karpenter, and the Descheduler.
> 

I don't think that NoAdmit solves this problem.  From the docs, NoAdmit is "NoSchedule" + don't start pods that are scheduled.  If it were "don't start pods only", maybe it could work but as it, it doesn't fully convey the concept of a temporary taint applied at startup that will be eventually removed.   Even if it were "don't start pods", I don't think it still fully implies that the taint is temporary.


> I'd hate to include this in karpenters API as a point in time workaround for something that will eventually be fixed upstream. One option is to support this taint effect in karpenter before it's merged upstream and do the upstream work in parallel.

The NoAdmit taint was introduced, commented out, in [2016](https://github.com/kubernetes/kubernetes/commit/52fb89ff7326b3570e9c42a435081dab559f610b) . That appears to be the initial implementation of taints and it hasn't been needed yet.  Agree that it doesn't seem to be a priority and we would need to do the work, but I also don't think it solves this particular problem.

To solve the problem, we would also need to treat 'NoAdmit' as a temporary taint by provisioning capacity for workloads that don't tolerate it and then assume something would remove the taint.  We would have implemented the concept of startup taints only hidden behind a magic effect name. If NoAdmit is ever implemented upstream, Karpenter would then be supporting it incorrectly as nothing in the NoAdmit docs indicate that it will always be removed.

File: pkg/controllers/provisioning/scheduling/inflightnode.go
Comment (tzneal) [reply pkg/controllers/provisioning/scheduling/inflightnode.go:82]: I didn't see a good way to accomplish it, there are quite a few differences.  I started to refactor out some of the common logic, but it didn't seem to help much.

File: charts/karpenter/crds/karpenter.sh_provisioners.yaml
Comment (njtran) [inline charts/karpenter/crds/karpenter.sh_provisioners.yaml:111]:
```suggestion
                  startup which are expected to be removed automatically within
```

File: pkg/apis/provisioning/v1alpha5/constraints.go
Comment (njtran) [inline pkg/apis/provisioning/v1alpha5/constraints.go:36]: Same thing here
```suggestion
	// StartupTaints are taints that are applied to nodes upon startup which are expected to be removed automatically
```

File: website/content/en/preview/tasks/provisioning.md
Comment (bwagner5) [inline website/content/en/preview/tasks/provisioning.md:86]:
```suggestion
Per the Cilium [docs](https://docs.cilium.io/en/stable/gettingstarted/taints/),  it's recommended to place a taint of `node.cilium.io/agent-not-ready=true:NoExecute` on nodes to allow Cilium to configure networking prior to other pods starting.  This can be accomplished via the use of Karpenter `startupTaints`.  These taints are placed on the node, but pods aren't required to tolerate these taints to be considered for provisioning.
```

File: pkg/apis/provisioning/v1alpha5/constraints.go
Comment (njtran) [inline pkg/apis/provisioning/v1alpha5/constraints.go:72]: nit: Is there a reason you don't want to inline this with the existing pattern? Is it to have the comment in there?

File: pkg/controllers/provisioning/scheduling/inflightnode.go
Comment (njtran) [inline pkg/controllers/provisioning/scheduling/inflightnode.go:20]: nit: can you move this down to group with the other github.com imports

File: pkg/apis/provisioning/v1alpha5/taints.go
Comment (njtran) [inline pkg/apis/provisioning/v1alpha5/taints.go:67]: Many tolerations can tolerate the same taint. Should we make the toleration just tolerate the key, not necessarily the value? We also just want NoScheduler here in all cases, right?

File: pkg/apis/provisioning/v1alpha5/util.go
Comment (njtran) [inline pkg/apis/provisioning/v1alpha5/util.go:27]: Is it possible the value of a taint would change? This would be out of spec but still have the same taint.

File: pkg/controllers/provisioning/provisioner.go
Comment (njtran) [inline pkg/controllers/provisioning/provisioner.go:212]: nit: inline?

File: pkg/apis/provisioning/v1alpha5/constraints.go
Comment (tzneal) [reply pkg/apis/provisioning/v1alpha5/constraints.go:72]: Not sure I follow, on line 83 I need to append again and I don't think this is valid syntax:

```go
taints := append(c.Taints, v1.Taint{Key: NotReadyTaintKey, Effect: v1.TaintEffectNoSchedule}, c.StartupTaints...)
```

**Comment by ellistarn** [comment]
> > I'm wary of the long term impact of including startup taints in the api. It's discussed in a number of places, but I'll restate here. Upstream has an unimplemented concept to address this use case (including the node restart case) called the NoAdmit taint effect. This allows scheduling but prevents the kubelet from running the pod, which is exactly what we need. One benefit of doing this upstream is that the change applies to all integrators like CAS, Karpenter, and the Descheduler.
> 
> I don't think that NoAdmit solves this problem. From the docs, NoAdmit is "NoSchedule" + don't start pods that are scheduled. If it were "don't start pods only", maybe it could work but as it, it doesn't fully convey the concept of a temporary taint applied at startup that will be eventually removed. Even if it were "don't start pods", I don't think it still fully implies that the taint is temporary.
> 

Think of these independently:
- NoSchedule: prevent scheduling
- NoExecute: evict any running pods
- NoAdmit: prevent the pod from running

In the initial implementation, there was some concept of combining the taint effects into a single effect, e.g. `NoScheduleNoAdmitNoExecute`. Wisely, upstream never went this direction and instead recommended that users specify both NoSchedule and NoExecute separately if both behaviors are desired. Similarly, I'd expect `NoAdmit` to behave this way, and don't expect that `NoScheduleNoAdmitNoExecute` would be accepted upstream. 

The idea that a taint exists at startup and will eventually be removed is superceded by the more general "don't run if this exists.". `NoAdmit` captures cases of startup and then removed, as well as flip flop (restart) scenarios. If we separate `NoScheduleNoAdmit` from `NoAdmit`, then we can just follow natural scheduling rules and the node will sort everything out at runtime. It's possible that the taint may take hours to be removed (think maintenance windows, or similar mechanisms), and neither karpenter nor the scheduler need to care.  

> > I'd hate to include this in karpenters API as a point in time workaround for something that will eventually be fixed upstream. One option is to support this taint effect in karpenter before it's merged upstream and do the upstream work in parallel.
> 
> The NoAdmit taint was introduced, commented out, in [2016](https://github.com/kubernetes/kubernetes/commit/52fb89ff7326b3570e9c42a435081dab559f610b) . That appears to be the initial implementation of taints and it hasn't been needed yet. Agree that it doesn't seem to be a priority and we would need to do the work, but I also don't think it solves this particular problem.
> 

See above.

> To solve the problem, we would also need to treat 'NoAdmit' as a temporary taint by provisioning capacity for workloads that don't tolerate it and then assume something would remove the taint. We would have implemented the concept of startup taints only hidden behind a magic effect name. If NoAdmit is ever implemented upstream, Karpenter would then be supporting it incorrectly as nothing in the NoAdmit docs indicate that it will always be removed.

Agree that a magical taint that is never implemented upstream could result in an awkward customer experience. I'm not sure whether or not it's worse than a special field to capture this edge case. There are are a couple paths here:

1. We implement NoAdmit, and successfully upstream the concept 🎉 
2. We implement NoAdmit, and upstream does something else -- may require deprecation in Karpenter
3. We implement startupTaints and upstream implements `NoAdmit` -- may require deprecation in Karpenter
4. We implement startupTaints and nothing happens upstream 🥈 .

How hard do you think it would be to get a directional sense from SIG Scheduling on this? My guess is that they would be supportive, since this concept is simple to implement, has precedence, and applies generally to CAS/Karpenter/etc.

**Comment by tzneal** [comment]
> Think of these independently:
> 
>     * NoSchedule: prevent scheduling
> 
>     * NoExecute: evict any running pods
> 
>     * NoAdmit: prevent the pod from running
> 

NoExecute does not just evict running pods though, it also implies NoSchedule.  I would expect NoAdmit to work similarly if it were to be implemented, though it could conceivably behave differently.

> 
> In the initial implementation, there was some concept of combining the taint effects into a single effect, e.g. `NoScheduleNoAdmitNoExecute`. Wisely, upstream never went this direction and instead recommended that users specify both NoSchedule and NoExecute separately if both behaviors are desired. Similarly, I'd expect `NoAdmit` to behave this way, and don't expect that `NoScheduleNoAdmitNoExecute` would be accepted upstream.

This isn't correct, if you put a NoExecute taint on a node it also prevents scheduling any pods unless they tolerate the taint.  If it didn't, the pod would just rapidly get scheduled and evicted immediately. 


> The idea that a taint exists at startup and will eventually be removed is superceded by the more general "don't run if this exists.". `NoAdmit` captures cases of startup and then removed, as well as flip flop (restart) scenarios. If we separate `NoScheduleNoAdmit` from `NoAdmit`, then we can just follow natural scheduling rules and the node will sort everything out at runtime. It's possible that the taint may take hours to be removed (think maintenance windows, or similar mechanisms), and neither karpenter nor the scheduler need to care.
> 

We need awareness of which taints are temporary and even if NoAdmit were to be accepted and implemented upstream, we can't assume that it would always be a temporary taint.  For in-flight nodes we care about the difference between a taint that we place on a node at startup which will exist for a short period of time vs a taint that can exist for hours.  In the latter case, we need to provision a new node.  In the first case, we would not.

The other issue is that for startup taints, we provision capacity even though the pod doesn't tolerate the taint because of the temporary nature of the taint.  This is different from other taints in which we only provision capacity if the pod tolerates the taint.

> Agree that a magical taint that is never implemented upstream could result in an awkward customer experience. I'm not sure whether or not it's worse than a special field to capture this edge case. There are are a couple paths here:
> 
>     1. We implement NoAdmit, and successfully upstream the concept 🎉


Assuming that:
- we can implement NoAdmit and get it merged upstream 
- it doesn't work as described in its initial docs and doesn't work similarly to the existing NoExecute and instead only prevents pods from running

Internally, we would have to add to 'NoAdmit' the concept of provisioning capacity for pods even if they don't tolerate the taint or it doesn't solve the current problem.   For in-flight nodes we have to treat NoAdmit as temporary to prevent us from provisioning excess capacity. If NoAdmit is then supported upstream, customers now can't taint their nodes with "NoAdmit" using Karpenter unless they want us to treat that with our special temporary taint handling. We then need to explain how/why we treat "NoAdmit" differently from every NoSchedule and NoExecute taints and are left in an awkward state.

>     2. We implement NoAdmit, and upstream does something else -- may require deprecation in Karpenter

We still have a special Karpenter only taint effect whose name doesn't clearly describe what it does.

>     3. We implement startupTaints and upstream implements `NoAdmit` -- may require deprecation in Karpenter

See 1, NoAdmit doesn't imply temporary so we still need the concept of a temporary taint.  We need something like startupTaints that lets customers express to Karpenter that a taint is temporary so pods don't need to tolerate it for the purposes of provisioning and so we can recognize it on in-flight nodes and know that the capacity will soon be available.  For the CCM and Cilium case, that capacity will be available faster than we can launch a node since every node goes through the same initialization process.

>     4. We implement startupTaints and nothing happens upstream 🥈 .

> How hard do you think it would be to get a directional sense from SIG Scheduling on this? My guess is that they would be supportive, since this concept is simple to implement, has precedence, and applies generally to CAS/Karpenter/etc.

I could see NoAdmit potentially solving some issues with pre-binding, but not all (e.g. extended resource zero'ing).  I don't think it can solve the 'temporary startup taints' problem at all.

File: pkg/apis/provisioning/v1alpha5/taints.go
Comment (tzneal) [reply pkg/apis/provisioning/v1alpha5/taints.go:67]: I wrote it this way so that if the node taint isn't exactly what we created, it doesn't get the special behavior.  I think that's the safest way to start this.  We place the taint on the node originally, so it will match.  If the taint then changes in some way, I don't think we can safely assume that we should still treat it as a startup taint.

File: pkg/apis/provisioning/v1alpha5/util.go
Comment (tzneal) [reply pkg/apis/provisioning/v1alpha5/util.go:27]: Same as above.

File: pkg/controllers/provisioning/provisioner.go
Comment (tzneal) [reply pkg/controllers/provisioning/provisioner.go:212]: I could go either way, but I think it's slightly less clear inlined.

**Comment by ellistarn** [comment]
I'll be offline due to a medical emergency. Don't block on this. 

You make some good points and I've learned a bunch from this discussion. However, I think you're missing the core of what I'm saying. Startup taints is a perfectly reasonable short term (and maybe long term) approach, even if I think we might be able to coordinate a few changes that would collapse the user facing complexity with a more general solution.

File: pkg/apis/provisioning/v1alpha5/constraints.go
Comment (njtran) [reply pkg/apis/provisioning/v1alpha5/constraints.go:72]: You're right.

**Review by bwagner5 — APPROVED** [review]
lgtm!

**Pull request merged**
